Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +25,75 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (currentNew_ < 0);
+}
+
+int BatteryStateOfChargeService::getRemainingHours() const
+{
+    return remainingHours_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return timeTillChargeOrDepletion_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(!timeOld.isNull())
+    {
+        double avgCurrent = (currentOld + currentNew_) / 2;
+        ampHours_ += (avgCurrent * timeOld.secsTo(timeNew_) / SECONDS_TO_HOURS);
+    }
+
+    //calculate remaining time for charge or depletion
+    double totalHours;
+    int hours;
+    int minutes;
+    int seconds;
+    int milliseconds;
+
+    if(isCharging())
+    {
+        totalHours = qAbs(totalAmpHoursUsed() / currentNew_);
+    }
+    else
+    {
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / currentNew_;
+    }
+
+    milliseconds = totalHours * MILLISECONDS_IN_HOUR;
+
+    hours = floor(totalHours);
+    milliseconds -= hours * MILLISECONDS_IN_HOUR;
+
+    minutes = floor(milliseconds / MILLISECONDS_IN_MINUTE);
+    milliseconds -= minutes * MILLISECONDS_IN_MINUTE;
+
+    seconds = floor(milliseconds / MILLISECONDS_IN_SECOND);
+    milliseconds -= seconds * MILLISECONDS_IN_SECOND;
+
+    if((hours - HOURS_IN_A_DAY) >= 0) // if 24 hours or more
+    {
+        remainingHours_ = hours - (HOURS_IN_A_DAY - 1);
```

Comment:
Why -1?
If you're at 30 total hours, shouldn't the remainingHours_ be 6? Instead of 30 - (24 - 1) = 7?

Reply:
The hours parameter in QTime will only hold up to 23 hours, not 24.

Reply:
I think this line deserves a comment then